src/frontend/Canonicalize.ml

```diff
@@ -165,30 +166,57 @@ let rec parens_stmt ({stmt; smeta} : typed_statement) : typed_statement =
           ; identifier
           ; initial_value= Option.map ~f:no_parens init
           ; is_global }
-    | While (e, s) -> While (no_parens e, parens_block s)
-    | IfThenElse (e, s1, Some ({stmt= IfThenElse _; _} as s2))
-     |IfThenElse (e, s1, Some {stmt= Block [({stmt= IfThenElse _; _} as s2)]; _})
-      ->
-        (* Flatten if ... else if ... constructs *)
-        IfThenElse (no_parens e, parens_block s1, Some (parens_stmt s2))
-    | IfThenElse (e, s1, s2) ->
-        IfThenElse (no_parens e, parens_block s1, Option.map ~f:parens_block s2)
     | For {loop_variable; lower_bound; upper_bound; loop_body} ->
         For
           { loop_variable
           ; lower_bound= keep_parens lower_bound
           ; upper_bound= keep_parens upper_bound
-          ; loop_body= parens_block loop_body }
+          ; loop_body= parens_stmt loop_body }
     | _ -> map_statement no_parens parens_stmt parens_lval ident stmt in
   {stmt; smeta}
 
-let repair_syntax program : untyped_program =
-  map_program
-    (repair_syntax_stmt (userdef_distributions program.functionblock))
+let rec blocks_stmt ({stmt; smeta} : typed_statement) : typed_statement =
+  let stmt_to_block ({stmt; smeta} : typed_statement) : typed_statement =
+    match stmt with
+    | Block _ -> blocks_stmt {stmt; smeta}
+    | _ ->
+        blocks_stmt
+        @@ mk_typed_statement
+             ~stmt:(Block [{stmt; smeta}])
+             ~return_type:smeta.return_type ~loc:smeta.loc in
+  let stmt =
+    match stmt with
+    | While (e, s) -> While (e, stmt_to_block s)
+    | IfThenElse (e, s1, Some ({stmt= IfThenElse _; _} as s2))
+     |IfThenElse (e, s1, Some {stmt= Block [({stmt= IfThenElse _; _} as s2)]; _})
+      ->
+        (* Flatten if ... else if ... constructs *)
+        IfThenElse (e, stmt_to_block s1, Some (blocks_stmt s2))
+    | IfThenElse (e, s1, s2) ->
+        IfThenElse (e, stmt_to_block s1, Option.map ~f:stmt_to_block s2)
+    | For ({loop_body; _} as f) ->
+        For {f with loop_body= stmt_to_block loop_body}
+    | _ -> map_statement ident blocks_stmt ident ident stmt in
+  {stmt; smeta}
+
+let repair_syntax program settings =
+  if settings.deprecations then
     program
+    |> map_program
+         (repair_syntax_stmt (userdef_distributions program.functionblock))
+  else program
```

Comment:
I'm inclined to say that `repair_syntax` should be enabled for all formatting, not just deprecations. It's not about syntax that's _going to be_ deprecated but syntax that _has already been_ deprecated.. A program that "needs" `repair_syntax` isn't even going to typecheck without it.

Reply:
At the moment, those programs _do_ fail to typecheck without `--print-canonical`. If we don't like that we could run it if and only if the formatter is activated, I suppose.